libshaderc_util/include/libshaderc_util/compiler.h

```diff
@@ -41,54 +42,17 @@ namespace shaderc_util {
 enum class PassId;
 
 // Initializes glslang on creation, and destroys it on completion.
-// This object is expected to be a singleton, so that internal
-// glslang state can be correctly handled.
-// TODO(awoloszyn): Once glslang no longer has static global mutable state
-//                  remove this class.
+// Used to tie gslang process operations to object lifetimes.
+// Additionally initialization/finalization of glslang is not thread safe, so
```

Comment:
I'm surprised.  I thought Glslang's  glslang::InitializeProcess and glslang::FinalizeProcess already did its own synchronization?
I'll have to go look.

Reply:
There is a lock created by InitializeProcess(), via ShInitialize(), that I think handles the thread safety guarantees for using glslang multiple times, but that lock doesn't look like it is controlling access during InitializeProcess() and FinalizeProcess(). It invokes ShInitialize() which has the warning comment:
> //
> // ShInitialize() should be called exactly once per process, not per thread.
> //

I was able to catch the race condition crash that the bots were seeing in the debugger, and the stack was in fillInKeywordMap(), which was invoked by ShInitialize().

My understanding is that setting up and tearing down glslang is still a single threaded operation, since there is global state that needs to be initialized, but invoking glslang can be concurrent since it only reads the global state.

With my change running the failing test 1000 times does not produce the test failure.

Reply:
Right.
Looking at ShFinalize I see a possibly different problem:  The lock is released after deciding whether to delete the data but before actually deleting the data.  :-/

Reply:
That sounds like something worth filing an issue against glslang for.